x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -161,14 +168,17 @@ public void clusterChanged(ClusterChangedEvent event) {
         final Version mappingVersion = oldestIndexMappingVersion(event.state());
         final ClusterHealthStatus indexStatus = indexMetaData == null ? null :
             new ClusterIndexHealth(indexMetaData, event.state().getRoutingTable().index(indexMetaData.getIndex())).getStatus();
+        // index name non-null iff state recovered
         final String concreteIndexName = indexMetaData == null ? INTERNAL_SECURITY_INDEX : indexMetaData.getIndex().getName();
         final State newState = new State(indexExists, isIndexUpToDate, indexAvailable, mappingIsUpToDate, mappingVersion, concreteIndexName,
             indexStatus);
         this.indexState = newState;
 
         if (newState.equals(previousState) == false) {
-            for (BiConsumer<State, State> listener : stateChangeListeners) {
-                listener.accept(previousState, newState);
+            // point in time iterator
+            final Iterator<BiConsumer<State, State>> stateListenerIterator = stateChangeListeners.iterator();
+            while (stateListenerIterator.hasNext()) {
+                stateListenerIterator.next().accept(previousState, newState);
```

Comment:
As far as I can tell this change doesn't actually do anything.
A `for` loop is just syntactic sugar over `iterator()` and `next()`, you haven't change the semantics. The `stateListenerIterator` is still backed by the same `List` and is at risk of `ConcurrentModificationException` issues if a listener is added/removed.

Assuming that's the problem you're trying to solve, then the options come down to:
1. `synchronize` whenever we use the list
2. switch the list implementation to `CopyOnWriteArrayList`
3. make a copy of the list before iterating over it

(2) is probably the best option. I don't think we add listeners very often, so making a list copy each time is probably OK, but you'd need to audit the places we modify the list.

Reply:
> As far as I can tell this change doesn't actually do anything.

It is a rewrite that unwraps the syntactic sugar because I find it confusing to use that when the list is concurrently modified. 

The list is already of the `CopyOnWriteArrayList` type.